**Table of contents**<a id='toc0_'></a>    
- 1. [Importing Data](#toc1_)    
- 2. [products.csv Data Consistency Checks](#toc2_)    
- 3. [orders_wrangled.csv Data Consistency Checks](#toc3_)    
- 4. [Exporting data frames as csv files](#toc4_)    

<!-- vscode-jupyter-toc-config
	numbering=true
	anchor=true
	flat=false
	minLevel=1
	maxLevel=6
	/vscode-jupyter-toc-config -->
<!-- THIS CELL WILL BE REPLACED ON TOC UPDATE. DO NOT WRITE YOUR TEXT IN THIS CELL -->

# 1. <a id='toc1_'></a>[Importing Data](#toc0_)

In [108]:
# Importing Libraries
import numpy as np
import pandas as pd
import os

In [109]:
Path = r'D:\Data Analysis\01-08-2025 Instacart Basket Analysis\Data'
df_pro = pd.read_csv(os.path.join(Path, 'Original Data', 'products.csv'), index_col=False)
df_ord = pd.read_csv(os.path.join(Path, 'Prepared Data', 'orders_wrangled.csv'), index_col=False)
df_ord.head()

Unnamed: 0  order_id  user_id eval_set  order_number  order_day_of_week  \
0           0   2539329        1    prior             1                  2   
1           1   2398795        1    prior             2                  3   
2           2    473747        1    prior             3                  3   
3           3   2254736        1    prior             4                  4   
4           4    431534        1    prior             5                  4   

   order_hour_of_day  days_since_prior_order  
0                  8                     NaN  
1                  7                    15.0  
2                 12                    21.0  
3                  7                    29.0  
4                 15                    28.0

In [110]:
# Creating a test dataframe to practice
df_test = pd.DataFrame()
df_test['mix'] = ['a', 'b', 1, True]
df_test

mix
0     a
1     b
2     1
3  True

# 2. <a id='toc2_'></a>[products.csv Data Consistency Checks](#toc0_)

In [111]:
# for-loop checking per column if dataframe contains any mixed-type columns. Weird checks wheter the datatype within columns are consistent and if stmt prints column names.
for col in df_test.columns.tolist():
  weird = (df_test[[col]].map(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_test[weird]) > 0:
    print (col)

mix


In [112]:
df_test['mix'] = df_test['mix'].astype('str')

In [113]:
#Figuring out which columns have null values
df_pro.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [ ]:
#Creating a new data frame that acts like a flag based on the missing value.
# ALt: df_proCleaned = df_pro.loc[df_pro['product_name'].isna() == True]
df_nan = df_pro[df_pro['product_name'].isnull() == True]
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

In [115]:
df_pro_clean = df_pro[df_pro['product_name'].isna() == False]
df_pro_clean.shape

(49677, 5)

In [116]:
df_pro_dups = df_pro_clean[df_pro_clean.duplicated()]
df_pro_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

In [117]:
df_pro_clean.shape

(49677, 5)

In [118]:
df_pro_clean_noDup = df_pro_clean.drop_duplicates()
df_pro_clean_noDup.shape

(49672, 5)

# 3. <a id='toc3_'></a>[orders_wrangled.csv Data Consistency Checks](#toc0_)

In [119]:
# Task 2
df_ord.describe()

Unnamed: 0      order_id       user_id  order_number  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710541e+06  1.710542e+06  1.029782e+05  1.715486e+01   
std    9.875817e+05  9.875817e+05  5.953372e+04  1.773316e+01   
min    0.000000e+00  1.000000e+00  1.000000e+00  1.000000e+00   
25%    8.552705e+05  8.552715e+05  5.139400e+04  5.000000e+00   
50%    1.710541e+06  1.710542e+06  1.026890e+05  1.100000e+01   
75%    2.565812e+06  2.565812e+06  1.543850e+05  2.300000e+01   
max    3.421082e+06  3.421083e+06  2.062090e+05  1.000000e+02   

       order_day_of_week  order_hour_of_day  days_since_prior_order  
count       3.421083e+06       3.421083e+06            3.214874e+06  
mean        2.776219e+00       1.345202e+01            1.111484e+01  
std         2.046829e+00       4.226088e+00            9.206737e+00  
min         0.000000e+00       0.000000e+00            0.000000e+00  
25%         1.000000e+00       1.000000e+01            4.000000e+00  
50%         3.000000e+00       1.300000e+01            7.000000e+00  
75%         5.000000e+00       1.600000e+01            1.500000e+01  
max         6.000000e+00       2.300000e+01            3.000000e+01

All columns, except for days_since_prior_order, show values that fall within their logical parameters. The lower values shown in the last column for count, mean could signify data consistency issues that will require further investigation.

In [120]:
# Task 3: checking for mixed-type data with for-loop did not output any column names: none found.
for col in df_test.columns.tolist():
  weird = (df_test[[col]].map(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_test[weird]) > 0:
    print (col)

In [121]:
#Task 5: Checking for missing values
df_ord.isna().sum()

Unnamed: 0                     0
order_id                       0
user_id                        0
eval_set                       0
order_number                   0
order_day_of_week              0
order_hour_of_day              0
days_since_prior_order    206209
dtype: int64

There are 206,209 order_ids with missing days_since_prior_order data which is not likely for it to be a data processing error in recognizing zero values since there are 67,755 order_id observations showing 0 for days lapsed. It could actually mean these were first time users with no order history. Imputing these values would alter our data frame patterns since both mean and median values will not represent real behavioral aspects of customers.

In [122]:
#Generating a subset that lists user_ids with missing/empty days since prior order values.
user_id_nan = df_ord.loc[df_ord['days_since_prior_order'].isna()]
user_id_nan

Unnamed: 0  order_id  user_id eval_set  order_number  \
0                 0   2539329        1    prior             1   
11               11   2168274        2    prior             1   
26               26   1374495        3    prior             1   
39               39   3343014        4    prior             1   
45               45   2717275        5    prior             1   
...             ...       ...      ...      ...           ...   
3420930     3420930    969311   206205    prior             1   
3420934     3420934   3189322   206206    prior             1   
3421002     3421002   2166133   206207    prior             1   
3421019     3421019   2227043   206208    prior             1   
3421069     3421069   3154581   206209    prior             1   

         order_day_of_week  order_hour_of_day  days_since_prior_order  
0                        2                  8                     NaN  
11                       2                 11                     NaN  
26                       1                 14                     NaN  
39                       6                 11                     NaN  
45                       3                 12                     NaN  
...                    ...                ...                     ...  
3420930                  4                 12                     NaN  
3420934                  3                 18                     NaN  
3421002                  6                 19                     NaN  
3421019                  1                 15                     NaN  
3421069                  3                 11                     NaN  

[206209 rows x 8 columns]

In [123]:
#Creating a new data frame with details of only user_ids detected by user_id_nan
user_id_withNAN = df_ord.loc[df_ord['user_id'].isin(user_id_nan['user_id'])].sort_values('user_id')
user_id_withNAN.head(50)

Unnamed: 0  order_id  user_id eval_set  order_number  order_day_of_week  \
0            0   2539329        1    prior             1                  2   
1            1   2398795        1    prior             2                  3   
6            6    550135        1    prior             7                  1   
5            5   3367565        1    prior             6                  2   
10          10   1187899        1    train            11                  4   
9            9   2550362        1    prior            10                  4   
8            8   2295261        1    prior             9                  1   
7            7   3108588        1    prior             8                  1   
3            3   2254736        1    prior             4                  4   
2            2    473747        1    prior             3                  3   
4            4    431534        1    prior             5                  4   
11          11   2168274        2    prior             1                  2   
23          23   3268552        2    prior            13                  4   
24          24    839880        2    prior            14                  3   
25          25   1492625        2    train            15                  1   
19          19   1718559        2    prior             9                  2   
13          13   1901567        2    prior             3                  1   
14          14    738281        2    prior             4                  2   
15          15   1673511        2    prior             5                  3   
16          16   1199898        2    prior             6                  2   
17          17   3194192        2    prior             7                  2   
18          18    788338        2    prior             8                  1   
20          20   1447487        2    prior            10                  1   
12          12   1501582        2    prior             2                  5   
22          22   3186735        2    prior            12                  1   
21          21   1402090        2    prior            11                  1   
36          36    521107        3    prior            11                  0   
35          35    676467        3    prior            10                  3   
38          38   2774568        3     test            13                  5   
37          37   1402502        3    prior            12                  1   
34          34   3160850        3    prior             9                  0   
33          33   3225766        3    prior             8                  0   
30          30   2710558        3    prior             5                  0   
29          29   2037211        3    prior             4                  2   
28          28   3002854        3    prior             3                  3   
27          27    444309        3    prior             2                  3   
32          32   1839752        3    prior             7                  0   
26          26   1374495        3    prior             1                  1   
31          31   1972919        3    prior             6                  0   
41          41    691089        4    prior             3                  4   
42          42     94891        4    prior             4                  5   
43          43   2557754        4    prior             5                  5   
44          44    329954        4     test             6                  3   
40          40   2030307        4    prior             2                  4   
39          39   3343014        4    prior             1                  6   
49          49   2196797        5    train             5                  0   
48          48    157374        5    prior             4                  1   
45          45   2717275        5    prior             1                  3   
46          46   1909121        5    prior             2                  0   
47          47   2267326        5    prior            

In [124]:
df_ord_clean = df_ord[df_ord['days_since_prior_order'].isna() == False]
df_ord_clean.shape

(3214874, 8)

Task 6 - The above results show that the null values are indeed due to inexistent order history for each user_id hence it would be better to not drop or impute and work with df_ord as it is. Created the df_ord_clean data frame without the 206209 df_ord NAN values in case we need to execute any special code that requires no nulll values and created a subset that lists the user_ids with missing days_since_prior_order values: namely user_id_nan created above.

In [125]:
#Task 7: Showing duplicate values
df_ord_dups = df_ord[df_ord.duplicated()]
df_ord_dups

Empty DataFrame
Columns: [Unnamed: 0, order_id, user_id, eval_set, order_number, order_day_of_week, order_hour_of_day, days_since_prior_order]
Index: []

Task 7. No duplicate values found in the df_ord data frame.

# 4. <a id='toc4_'></a>[Exporting data frames as csv files](#toc0_)

In [126]:
# Task 9 - Exporting df_pro_clean_noDup and df_ord_clean data as csv files and may use df_ord depending on requirements.
df_pro_clean_noDup.to_csv(os.path.join(Path, 'Prepared Data', 'df_pro_NoNANDups.csv'))
df_ord_clean.to_csv(os.path.join(Path, 'Prepared Data', 'df_ord_NoNANDups.csv'))